### DISCUSSION TITLE
Could fipy load the .msh file from the gmsh or build a natural river mesh file with Gmsh2D?

### DISCUSSION THREAD

**Discussion by xrc123456** [usnistgov/fipy#976] [Q&A]
When I want to load the .msh file which have been build by the gmsh, it will have many issues, like the file have been occupied. 
When I want to build a natural 2D mesh byu Gmsh2D , the fipy can't build it, but for a rectangle data, it could do. But we want to simulate something in a river , it almost not the regular shape. Hope you can help us to solve this problem, Thanks...

```python
# Define the characteristic length
lc = 0.2  # Adjust this value as needed for your mesh

data_points = [
    (1, 2),
    (1, 3),
    (2, 2.5),
    (2, 3.5),
    (3, 1.5),
    (3, 2.5)
]

point_str = ""
line_str = ""

for i, point in enumerate(data_points):
    # Include lc in each point definition
    point_str += "Point(%d) = {%f, %f, 0, %f};\n" % (i+1, point[0], point[1], lc)

    if i > 0:
        line_str += "Line(%d) = {%d, %d};\n" % (i, i, i+1)

# Closing the loop
line_str += "Line(%d) = {%d, 1};\n" % (len(data_points), len(data_points))

# Creating a loop and a surface
loop_str = "Line Loop(%d) = {%s};\n" % (len(data_points)+1, ", ".join([str(i) for i in range(1, len(data_points)+1)]))
surface_str = "Plane Surface(%d) = {%d};\n" % (len(data_points)+2, len(data_points)+1)

# Complete Gmsh script
gmsh_script = point_str + line_str + loop_str + surface_str

# Creating mesh in FiPy
from fipy import Gmsh2D

mesh = Gmsh2D(gmsh_script)

# Visualization
from fipy import Viewer

viewer = Viewer(mesh)
viewer.plot()
```

**Comment by guyer** [comment]
The boundary you have defined crosses itself. That's not valid in Gmsh. This is not a FiPy issue.

<img width="614" alt="image" src="https://github.com/usnistgov/fipy/assets/1830725/3de9322a-48de-423e-9364-1ad4a138d7c1">

**Reply by xrc123456** [reply]
Yes, I found this problem now....really appreciate it.....We thought we must use the gmsh and then transfer the msh to the fipy from the website, but today I think maybe the Gmsh fron Fipy also very good to do this work. 

![image](https://github.com/usnistgov/fipy/assets/74932112/f7eacb7c-ae94-4e2f-a3c4-653f7da1aaf2)

it is very good results, BUT I want to confirm that could we use Fipy to calculate the N-S and CFD methods like Openform and control the boundary and time? Also which package should be better match with Fipy that let users can configure the condition in a visual platform. I think the figure you sent me before is used a professional one, right?
